keep/api/core/db.py

```diff
@@ -3248,6 +3248,71 @@ def remove_alerts_to_incident_by_incident_id(
         return deleted
 
 
+class DestinationIncidentNotFound(Exception):
+    pass
+
+
+def merge_incidents_to_id(
+    tenant_id: str,
+    source_incident_ids: List[UUID],
+    destination_incident_id: UUID,
+    merged_by: str | None = None,
+) -> Optional[Incident]:
+    with Session(engine) as session:
+        destination_incident = session.exec(
+            select(Incident)
+            .where(
+                Incident.tenant_id == tenant_id, Incident.id == destination_incident_id
+            )
+            .options(joinedload(Incident.alerts))
+        ).first()
+
+        if not destination_incident:
+            # TODO: maybe allow to create a new incident if the destination incident does not exist
+            raise DestinationIncidentNotFound(
+                f"Destination incident with id {destination_incident_id} not found"
+            )
+
+        source_incidents = session.exec(
+            select(Incident).filter(
+                Incident.tenant_id == tenant_id,
+                Incident.id.in_(source_incident_ids),
+            )
+        ).all()
+
+        alerts_to_add_ids = []
+        for source_incident in source_incidents:
+            alerts_to_add_ids.extend([alert.id for alert in source_incident.alerts])
+            source_incident.merged_into_incident_id = destination_incident.id
+            source_incident.merged_at = datetime.now(tz=timezone.utc)
+            source_incident.status = IncidentStatus.MERGED.value
+            source_incident.merged_by = merged_by
+            try:
+                # TODO: optimize, process in bulk
+                remove_alerts_to_incident_by_incident_id(
+                    tenant_id,
+                    source_incident.id,
+                    [alert.id for alert in source_incident.alerts],
+                )
+            except OperationalError as e:
+                logger.error(
+                    f"Error removing alerts to incident {source_incident.id}: {e}"
+                )
+
+        try:
+            add_alerts_to_incident(
```

Comment:
I would probably suggest moving add_alerts_to_incident close to remove_alerts_to_incident_by_incident_id. So it will remove and immediately add... Per-incident. In this case exception in add_alerts_to_incident will damage only one Incident, not all.